streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala

```diff
@@ -190,283 +281,187 @@ class WriteAheadLogSuite extends SparkFunSuite with BeforeAndAfter {
+    f.onComplete {
+      case Success(v) =>
+        assert(v === walHandle) // return our mock handle after the write
+        if (numSuccess != null) numSuccess.incrementAndGet()
+      case Failure(v) => if (numFail != null) numFail.incrementAndGet()
+    }(walBatchingThreadPool)
   }
 
   /**
-   * Write data to a file using the writer class and return an array of the file segments written.
+   * In order to block the writes on the writer thread, we mock the write method, and block it
+   * for some time with a promise.
    */
-  def writeDataUsingWriter(
-      filePath: String,
-      data: Seq[String]
-    ): Seq[FileBasedWriteAheadLogSegment] = {
-    val writer = new FileBasedWriteAheadLogWriter(filePath, hadoopConf)
-    val segments = data.map {
-      item => writer.write(item)
-    }
-    writer.close()
-    segments
+  private def writeBlockingPromise(wal: WriteAheadLog): Promise[Any] = {
+    // we would like to block the write so that we can queue requests
+    val promise = Promise[Any]()
+    when(wal.write(any[ByteBuffer], any[Long])).thenAnswer(
+      new Answer[FileBasedWriteAheadLogSegment] {
+        override def answer(invocation: InvocationOnMock): FileBasedWriteAheadLogSegment = {
+          Await.ready(promise.future, 4.seconds)
+          walHandle
+        }
+      }
+    )
+    promise
   }
 
-  /** Write data to rotating files in log directory using the WriteAheadLog class. */
-  def writeDataUsingWriteAheadLog(
-      logDirectory: String,
-      data: Seq[String],
-      manualClock: ManualClock = new ManualClock,
-      closeLog: Boolean = true,
-      clockAdvanceTime: Int = 500,
-      closeFileAfterWrite: Boolean = false): FileBasedWriteAheadLog = {
-    if (manualClock.getTimeMillis() < 100000) manualClock.setTime(10000)
-    val wal = new FileBasedWriteAheadLog(new SparkConf(), logDirectory, hadoopConf, 1, 1,
-      closeFileAfterWrite)
-
-    // Ensure that 500 does not get sorted after 2000, so put a high base value.
-    data.foreach { item =>
-      manualClock.advance(clockAdvanceTime)
-      wal.write(item, manualClock.getTimeMillis())
+  test("BatchedWriteAheadLog - records get added to a queue") {
+    val numSuccess = new AtomicInteger()
+    val numFail = new AtomicInteger()
+    val wal = new BatchedWriteAheadLog(fileBasedWAL)
+
+    val promise = writeBlockingPromise(fileBasedWAL)
+
+    // make sure queue is empty initially
+    assert(wal.getQueueLength === 0)
+    val event1 = "hello"
+    val event2 = "world"
+    val event3 = "this"
+    val event4 = "is"
+    val event5 = "doge"
+
+    eventFuture(wal, event1, 5L, numSuccess, numFail)
+    eventFuture(wal, event2, 10L, numSuccess, numFail)
+    eventFuture(wal, event3, 11L, numSuccess, numFail)
+    eventFuture(wal, event4, 12L, numSuccess, numFail)
+    eventFuture(wal, event5, 20L, numSuccess, numFail)
+
+    eventually(Eventually.timeout(2 seconds)) {
+      // the first element will immediately be taken and the rest will get queued
```

Comment:
This is not guaranteed. `BatchedWriteAheadLog.flushRecords` just drains all records in the queue, and the queue length could be anything. You may not be able to observe 4.